docs/telepathy-sofiasip.8.in
.TH TELEPATHY-SOFIASIP "8" "February 2010" "Telepathy" "D-Bus services"
\" This man page was written by Simon McVittie for the Debian project,
\" but may be used by others.
\" Copyright © 2007-2010 Collabora Ltd. <http://www.collabora.co.uk/>
\" It may be distributed under the same terms as telepathy-sofiasip itself.
.SH NAME
telepathy-sofiasip \- Telepathy connection manager for SIP
.SH SYNOPSIS
\fB@libexecdir@/telepathy\-sofiasip\fR
.SH DESCRIPTION
\fBtelepathy\-sofiasip\fR implements the Telepathy D-Bus specification for
the Session Initiation Protocol, allowing Telepathy clients like
.BR empathy (1)
to connect to Voice over IP providers that implement SIP.
.PP
It is a D-Bus service which runs on the session bus, and should usually be
started automatically by D-Bus activation. However, it might be useful to
start it manually for debugging.
.SH OPTIONS
There are no command-line options.
.SH ENVIRONMENT
.TP
\fBTPSIP_DEBUG\fR=\fItype\fR
May be set to "all" for full debug output, or various undocumented options
(which may change from release to release) to filter the output.
.TP
\fBTPSIP_LOGFILE\fR=\fIfilename\fR
May be set to the name of a file to overwrite with log messages. The filename
may be prefixed with \fB+\fR to append messages to an existing file instead
of overwriting it.
.TP
\fBTPSIP_PERSIST\fR
May be set to any value to avoid telepathy-sofiasip's usual automatic exit
when there have been no connections for a few seconds.
.TP
\fBTPORT_LOG\fR
May be set to any value to print all parsed SIP messages at the transport
layer (this functionality is provided by the underlying Sofia-SIP library,
which also offers various more specialized debugging options).
.SH SEE ALSO
.IR http://telepathy.freedesktop.org/ ,
.BR empathy (1)


NEWS
===========================
NEWS for telepathy-sofiasip
===========================

List of changes in each public release of telepathy-sofiasip (most 
recent at the top).

See also ChangeLog.

Changes in release 0.6.0 (2010-02-15)
-------------------------------------

- Supports the new StreamHandler interface method CodecsUpdated.
- Revisited code controlling stream directionality and pending flags
  (fd.o #26070).
- Made critical warnings fatal again, after telepathy-glib stopped doing so.
  (fd.o #26153).
- Try to disable locally originated re-INVITES as much as possible
  when 'immutable-streams' parameter is true (fd.o #26387).
- Implemented the Debug interface (fd.o #26388).
- Implemented the DBus properties of Media.StreamHandler (fd.o #26384).
- Build fixes and improvements.
- Support configure option --enable-silent-rules if using Automake >= 1.11
- Implemented properties InitialAudio, InitialVideo, and ImmutableStreams.
- Introduced the connection manager parameter 'immutable-streams' to
  disable stream list changes mid-call, as well as hold.
- Pass the ptime and maxptime attributes from remote SDP session and media
  descriptions as codec parameters. Values from a=fmtp are also considered. 

Changes in release 0.5.19 (2009-12-02)
--------------------------------------

- Implemented proper codec parameter formatting for telephone/event
  payloads, and the supporting infrastructure for more formats.
  (fd.o bug #24615)
- Ignore the trivial alias for self handle that is produced from the
  SIP URI.
- Implemented call state In_Progress.
- Properly flag password parameters as secret, also in the .manager file
  (fd.o bug #24561).
- Autotools cleanups.
- Normalize the alias for self handle, and properly format it into a
  double-quoted string for the SIP From header.
- Use heartbeat more carefully, with per-connection wakeups timed
  in accordance with actual keepalive intervals.
- Eliminate race conditions in hold/unhold code.

Changes in release 0.5.18 (2009-08-27)
--------------------------------------

- Don't delay the final 200 response to incoming messages.
- Use sips: scheme in URI normalization when TLS transport
  is selected.
- Treat 183 Session Progress response with setting the Ringing call state.

Changes in release 0.5.17 (2009-07-25)
--------------------------------------

- Do not check an unwarranted handle assertion (fd.o bug #17751).
- Optionally using iphb on Maemo as a trigger for keepalives.

Changes in release 0.5.16 (2009-04-14)
--------------------------------------

- Updated to Telepathy specification 0.17.22:
  always allow self handle to leave Group for call channels
  (fd.o bug #20733). Correspondingly, bumped the libtelepathy-glib minimal
  version to 0.7.27.
- Emit MembersChanged with proper actor on call channels terminated locally.
- Map the 404 response status to member removal reason NoAnswer
  if the call peer has been contacted.
- Don't destroy the media session object before closing the channel
  on dispose.

Changes in release 0.5.15 (2009-02-10)
--------------------------------------

- Use the error code Invalid_Handle for failed handle normalization,
  as per recent revisions of telepathy-spec (fd.o bug #19612).
- Ignore provisional responses to a MESSAGE.
- Handle an incoming BYE with a proper media channel group change.
- Restored the backwards compatible behavior of AddMembers on a media
  channel (fd.o bug #19652).
- Updated the coverage instrumentation support, with lcov reports,
  taken from telepathy-gabble.
- Fixed issues reported by Coverity static analysis checks.

Changes in release 0.5.14 (2009-01-13)
--------------------------------------

- Bumped up the Sofia-SIP requirement to 1.12.10, to avoid a crasher bug
  in earlier versions.
- Zero-initialize the GValue for the alias pair.

Changes in release 0.5.13 (2008-11-28)
--------------------------------------

- Fixed a few brown paper bag bugs with the new interfaces and properties
  on the media channel.
- Do advertise Requests interface as supported by the connection.

Changes in release 0.5.12 (2008-11-26)
--------------------------------------

- Implemented Telepathy spec 0.17.15, compatible with 0.17.14.
- Implemented the Requests interface on connections.
  The factories have been switched to implement TpChannelManager. 

Changes in release 0.5.11 (2008-10-30)
--------------------------------------

- Migrated the source repository to Git.
- Updated to telepathy spec 0.17.13.
- Improved handle normalization.
- Implemented the Aliasing interface for connections, returning a
  simplified rendering of remote SIP URIs. The alias for the self handle
  is now settable and transfers to the display name in the From: header.
- The signal MembersChanged in call channels now follows protocol events
  more accurately.
- Removed the hardcoded timeout for outgoing INVITE requests.
- The final 200 response for an incoming MESSAGE now follows Telepathy
  acknowledgement for the message.
- A Text channel is reopened if closed by the client while still having
  unacknowledged messages.
- Depends on glib 2.16 or later.
- Depends on libtelepathy-glib 0.7.17 or later.

Changes in release 0.5.10 (2008-07-15)
--------------------------------------

- Implemented glare resolution in case of a 491 Request Pending response.
- Implemented D-Bus properties for Channel and Group implementations.
- Rewritten the parser for a:fmtp SDP attribute. Fixes fd.o bug #16405

Changes in release 0.5.9 (2008-05-21)
-------------------------------------

- Avoid redundant pushes of remote codec lists to Telepathy.
- Revamped test suite synced from Gabble.
- Full support for RTCP stream components.
- Use telepathy-glib 0.7.8 for Hold interface helpers
- Make sure the call channel is always closed after a remote error
  response is received and signalled as peer member removal.
- Fixed a regression in handling newly requested streams during an ongoing
  negotiation.

src/telepathy-sofiasip.c
/*
 * sip-connection.c - Source for TpsipConnection
 * Copyright (C) 2005 Collabora Ltd.
 * Copyright (C) 2005-2008, 2010 Nokia Corporation
 *   @author Kai Vehmanen <first.surname@nokia.com>
 *   @author Martti Mela <first.surname@nokia.com>
 *   @author Mikhail Zabaluev <mikhail.zabaluev@nokia.com>
 *
 * Based on telepathy-gabble implementation (gabble-connection).
 *   @author See gabble.c
 *
 * This work is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This work is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>
#include "debug.h"

#include "sip-connection-manager.h"
#include <telepathy-glib/run.h>
#include <telepathy-glib/debug.h>

#include <sofia-sip/su_log.h>

static TpBaseConnectionManager *
construct_cm (void)
{
  return (TpBaseConnectionManager *)g_object_new (
      TPSIP_TYPE_CONNECTION_MANAGER, NULL);
}

static void
sofia_log_handler (void *logdata, const char *format, va_list args)
{
  GString *buf = (GString *)logdata;
  gsize pos;
  gsize bytes_available;
  gsize length_added;

  g_assert (buf != NULL);

  /* Append the formatted message at the end of the buffer */
  pos = buf->len;
  for (;;)
    {
      bytes_available = buf->allocated_len - pos;
      length_added = g_vsnprintf (buf->str + pos,
                                  bytes_available,
                                  format, args);
      if (length_added < bytes_available)
        {
          buf->len = pos + length_added;
          g_assert (!buf->str[buf->len]);
          break;
        }
      g_string_set_size (buf, pos + length_added);
    }

  /* If we have a terminated line, log it */
  if (buf->str[buf->len - 1] == '\n')
    {
      tpsip_log (TPSIP_DEBUG_SOFIA, G_LOG_LEVEL_DEBUG, "%s", buf->str);
      g_string_truncate (buf, 0);
    }
}

static gpointer
sofia_log_init ()
{
  GString *buf;

  buf = g_string_sized_new (80);

  su_log_redirect (NULL, sofia_log_handler, buf);

  return buf;
}

static void
sofia_log_finalize (gpointer logdata)
{
  GString *buf = (GString *)logdata;

  if (buf->len != 0)
    {
      /* Don't use tpsip_log here because the CM has already been finalized, so
       * out TpDebugSender will have too. It isn't crucial, anyway. */
      g_debug ("%s", buf->str);
      g_message ("last Sofia log message was not newline-terminated");
    }

  g_string_free (buf, TRUE);
}

int
main (int argc, char** argv)
{
  int status;
  gpointer logdata;
  guint fatal_mask;

  fatal_mask = g_log_set_always_fatal (G_LOG_FATAL_MASK);
  fatal_mask |= G_LOG_LEVEL_CRITICAL;
  g_log_set_always_fatal (fatal_mask);  

#ifdef ENABLE_DEBUG
  tpsip_debug_set_flags_from_env ();

  if (g_getenv ("TPSIP_PERSIST") || g_getenv ("SOFIASIP_PERSIST"))
    {
      tp_debug_set_persistent (TRUE);
    }
#endif

  tp_debug_divert_messages (g_getenv ("TPSIP_LOGFILE"));

  logdata = sofia_log_init ();

  status = tp_run_connection_manager ("telepathy-sofiasip", VERSION,
                                      construct_cm, argc, argv);

  sofia_log_finalize (logdata);

  return status;
}
